fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -42,50 +61,148 @@ impl<T> Interpreter<PredicateStorage, T> {
+                            return Err(PredicateVerificationFailed::GasNotSpecified);
+                        };
 
-            if !matches!(vm.verify_predicate()?, ProgramState::Return(0x01)) {
-                return Err(PredicateVerificationFailed::False);
+                        vm.init_predicate(context, *tx, balances.clone(), available_gas)?;
+                        available_gas
+                    }
+                    PredicateRunKind::Estimating(tx) => {
+                        let context = Context::PredicateEstimation { program: predicate };
+                        let tx_available_gas = params
+                            .max_gas_per_tx
+                            .checked_sub(cumulative_gas_used)
+                            .ok_or_else(|| Bug::new(BugId::ID003, GlobalGasUnderflow))?;
+                        let available_gas = core::cmp::min(params.max_gas_per_predicate, tx_available_gas);
+
+                        vm.init_predicate(context, *tx, balances.clone(), available_gas)?;
+                        available_gas
+                    }
+                };
+
+                let result = vm.verify_predicate();
+                let is_successful = matches!(result, Ok(ProgramState::Return(0x01)));
+
+                let gas_used = available_gas
+                    .checked_sub(vm.remaining_gas())
+                    .ok_or_else(|| Bug::new(BugId::ID004, GlobalGasUnderflow))?;
+                cumulative_gas_used = cumulative_gas_used
+                    .checked_add(gas_used)
+                    .ok_or_else(|| PredicateVerificationFailed::OutOfGas)?;
+
+                match &mut kind {
+                    PredicateRunKind::Verifying(_) => {
+                        if !is_successful {
+                            result?;
+                            return Err(PredicateVerificationFailed::False);
+                        }
+
+                        if vm.remaining_gas() != 0 {
+                            return Err(PredicateVerificationFailed::GasMismatch);
+                        }
+                    }
+                    PredicateRunKind::Estimating(tx) => match &mut tx.inputs_mut()[i] {
+                        Input::CoinPredicate(CoinPredicate { predicate_gas_used, .. })
+                        | Input::MessageCoinPredicate(MessageCoinPredicate { predicate_gas_used, .. })
+                        | Input::MessageDataPredicate(MessageDataPredicate { predicate_gas_used, .. }) => {
+                            *predicate_gas_used = gas_used;
+                        }
+                        _ => {
+                            unreachable!("It was checked before during iteration over predicates")
+                        }
+                    },
+                }
             }
+        }
 
-            remaining_gas = vm.registers[RegId::GGAS];
+        match kind {
+            PredicateRunKind::Verifying(tx) => {
+                if cumulative_gas_used > tx.limit() {
```

Comment:
Shouldn't `cumulative_gas_used` also factor in the min_fee? Otherwise this limit check is only checking the total gas used by predicates against the total limit for the whole transaction.